### PR TITLE
Important: #95647 - Composer installations and extension usage

### PR DESCRIPTION
**Pull request by linawolf**
https://docs.typo3.org/c/typo3/cms-core/master/en-us/Changelog/11.4/Important-95647-ComposerInstallationsAndExtensionUsage.html
refs https://github.com/TYPO3-Documentation/TYPO3CMS-Reference-CoreApi/issues/1449

releases: master, 11.5

### CODE REVIEW

File: Documentation/ApiOverview/Services/Developer/Implementing.rst [hunk 1]
```diff
@@ -13,9 +13,9 @@ Implementing a service
 There are no tools to get you started coding a new service.
 However there is not much that needs to be done.
 
-A service should be packaged into an extension. This means that you
-will need at least a declaration file :file:`ext_emconf.php` and
-an extension's icon. The class file for your service should be
+A service should be packaged into an extension. Read chapter
```
Comment (mbrodala) [inline Documentation/ApiOverview/Services/Developer/Implementing.rst:16]:
```suggestion
A service should be packaged into an extension. The chapter
```

File: Documentation/ApiOverview/Services/Developer/Implementing.rst [hunk 2]
```diff
@@ -13,9 +13,9 @@ Implementing a service
...
+:ref:`Files and locations <extension-files-locations>` about minimal
```
Comment (mbrodala) [inline Documentation/ApiOverview/Services/Developer/Implementing.rst:17]:
```suggestion
:ref:`Files and locations <extension-files-locations>` explains the minimal
```

File: Documentation/ApiOverview/Configuration/ConfigurationOverview.rst [hunk 3]
```diff
@@ -43,10 +43,10 @@ Extension files
 ---------------
 
 :ref:`composer.json <composer-json>`
-   Composer configuration
+   Composer configuration, required in composer-based installations
```
Comment (mbrodala) [inline Documentation/ApiOverview/Configuration/ConfigurationOverview.rst:46]:
```suggestion
   Composer configuration, required in Composer installations
```

File: Documentation/ApiOverview/Configuration/ConfigurationOverview.rst [hunk 4]
```diff
@@ -43,10 +43,10 @@ Extension files
...
 
 :ref:`ext_emconf.php <extension-declaration>`
-   (required) Extension declaration
+   Extension declaration, required in legacy installations
```
Comment (mbrodala) [inline Documentation/ApiOverview/Configuration/ConfigurationOverview.rst:49]:
```suggestion
   Extension declaration, required in Legacy installations
```

File: Documentation/ExtensionArchitecture/ComposerJson/Index.rst [hunk 5]
```diff
@@ -9,13 +9,14 @@
 composer.json
 =============
 
-*-- required*
+*-- required* in composer-based installations
```
Comment (mbrodala) [inline Documentation/ExtensionArchitecture/ComposerJson/Index.rst:12]:
```suggestion
*-- required* in Composer installations
```

File: Documentation/ExtensionArchitecture/ComposerJson/Index.rst [hunk 6]
```diff
@@ -9,13 +9,14 @@
...
 
 .. note::
 
    While the file :file:`composer.json` is currently not strictly required
-   for an extension to function properly, it is considered
-   bad practice not to add one. That is why we classify it as "required".
+   for an extension to function properly in legacy non-Composer installations
```
Comment (mbrodala) [inline Documentation/ExtensionArchitecture/ComposerJson/Index.rst:17]:
```suggestion
   for an extension to function properly in Legacy installations
```

File: Documentation/ExtensionArchitecture/ComposerJson/Index.rst [hunk 7]
```diff
@@ -75,6 +76,18 @@ Subsequently:
   general Composer information
 * see :ref:`ext-composer-json-properties` below for TYPO3 specific hints
 
+.. versionchanged:: 11.4
+   The ordering of installed extensions and their dependencies are loaded from
+   the :file:`composer.json` file, instead of :file:`ext_emconf.php` in
+   composer-based installations.
```
Comment (mbrodala) [inline Documentation/ExtensionArchitecture/ComposerJson/Index.rst:82]:
```suggestion
   Composer installations.
```

File: Documentation/ExtensionArchitecture/ComposerJson/Index.rst [hunk 8]
```diff
@@ -187,9 +200,23 @@ require
 (*required*)
 
 At the least, you will want to require `typo3/cms-core`.
-You can add other system extensions and third party extensions,
+You should add other system extensions and third party extensions,
```
Comment (mbrodala) [inline Documentation/ExtensionArchitecture/ComposerJson/Index.rst:203]:
```suggestion
You must add other system extensions and third party extensions,
```

You'll get errors if you don't do this.

File: Documentation/ExtensionArchitecture/ComposerJson/Index.rst [hunk 9]
```diff
@@ -187,9 +200,23 @@ require
...
 if your extension depends on them.
 
+In Composer-based installations the loading order of extensions and their
```
Comment (mbrodala) [inline Documentation/ExtensionArchitecture/ComposerJson/Index.rst:206]:
```suggestion
In Composer installations the loading order of extensions and their
```

File: Documentation/ExtensionArchitecture/ComposerJson/Index.rst [hunk 10]
```diff
@@ -187,9 +200,23 @@ require
...
+dependencies is derived from :json:`require` and :json:`suggest`
+
+
+
+suggest
+-------
+
+You should add other system extensions and third party extensions,
+if your extension depends on them in case they are loaded.
```
Comment (mbrodala) [inline Documentation/ExtensionArchitecture/ComposerJson/Index.rst:215]:
```suggestion
if your extension has an optional dependency on them.
```

File: Documentation/ExtensionArchitecture/ComposerJson/Index.rst [hunk 11]
```diff
@@ -187,9 +200,23 @@ require
...
+
+In Composer-based installations the loading order of extensions and their
```
Comment (mbrodala) [inline Documentation/ExtensionArchitecture/ComposerJson/Index.rst:217]:
```suggestion
In Composer installations the loading order of extensions and their
```

File: Documentation/ExtensionArchitecture/DeclarationFile/Index.rst [hunk 12]
```diff
@@ -6,20 +6,28 @@
...
 
-=================================
-Declaration File (ext_emconf.php)
-=================================
+=========================================
+Declaration file (:file:`ext_emconf.php`)
+=========================================
 
-*-- required*
+*-- required*  in legacy installations
```
Comment (mbrodala) [inline Documentation/ExtensionArchitecture/DeclarationFile/Index.rst:13]:
```suggestion
*-- required*  in Legacy installations
```

File: Documentation/ExtensionArchitecture/DeclarationFile/Index.rst [hunk 13]
```diff
@@ -6,20 +6,28 @@
...
 
-The :file:`ext_emconf.php` is the single most important file in an extension.
-Without it, the Extension Manager (EM) will not detect the extension, much less
-be able to install it. This file contains a declaration of what the extension
-is or does for the EM. The only thing included
+The :file:`ext_emconf.php` is used in legacy non-Composer installations to
```
Comment (mbrodala) [inline Documentation/ExtensionArchitecture/DeclarationFile/Index.rst:15]:
```suggestion
The :file:`ext_emconf.php` is used in Legacy installations to
```

File: Documentation/ExtensionArchitecture/DeclarationFile/Index.rst [hunk 14]
```diff
@@ -6,20 +6,28 @@
...
+supply information about the extension to the :guilabel:`Extension Manager`. In
```
Comment (mbrodala) [inline Documentation/ExtensionArchitecture/DeclarationFile/Index.rst:16]:
```suggestion
supply information about the extension in the :guilabel:`Extensions` module. In
```

There is no *Extension Manager* label in the TYPO3 backend, only the *Extensions* module.

File: Documentation/ExtensionArchitecture/DeclarationFile/Index.rst [hunk 15]
```diff
@@ -6,20 +6,28 @@
...
+these installations the ordering of installed extensions and their dependencies
+are loaded from this file as well.
```
Comment (mbrodala) [inline Documentation/ExtensionArchitecture/DeclarationFile/Index.rst:18]:
```suggestion
depends on this file as well.
```

File: Documentation/ExtensionArchitecture/DeclarationFile/Index.rst [hunk 16]
```diff
@@ -6,20 +6,28 @@
...
+
+.. versionchanged:: 11.4
+   The ordering of installed extensions and their dependencies are loaded from
+   the :file:`composer.json` file, instead of :file:`ext_emconf.php` in
+   composer-based installations.
```
Comment (mbrodala) [inline Documentation/ExtensionArchitecture/DeclarationFile/Index.rst:23]:
```suggestion
   Composer installations.
```

File: Documentation/ExtensionArchitecture/DeclarationFile/Index.rst [hunk 17]
```diff
@@ -6,20 +6,28 @@
...
+
+The only thing included
 is an associative array, :php:`$EM_CONF[extension key]`.
```
Comment (mbrodala) [inline Documentation/ExtensionArchitecture/DeclarationFile/Index.rst:26]:
```suggestion
is an associative array, :php:`$EM_CONF[$_EXTKEY]`.
```

File: Documentation/ExtensionArchitecture/DeclarationFile/Index.rst [hunk 18]
```diff
@@ -6,20 +6,28 @@
...
 The keys are described in the table below.
 
-This file is overwritten, when extensions are imported from the online repository. So don't write your custom code in this file - only change
+This file is overwritten, when extensions are imported from the online
+repository. So don't write your custom code in this file - only change
```
Comment (mbrodala) [inline Documentation/ExtensionArchitecture/DeclarationFile/Index.rst:30]:
```suggestion
repository (TER). So don't write your custom code in this file - only change
```

File: Documentation/ExtensionArchitecture/DeclarationFile/Index.rst [hunk 19]
```diff
@@ -169,16 +177,28 @@ $_EXTKEY is set globally and contains the extension key.
...
          corresponding branch, thus ``.99``). Also the extension has been
-         tested and is known to work properly with PHP 7.2, 7.3 and 7.4. It
+         tested and is known to work properly with PHP 7.4. and 8.0 It
          will conflict with "templavoilaplus" (any version) and it is suggested
-         that it might be worth installing "news" (version at least 7.3.0).
+         that it might be worth installing "news" (version at least 9.0.0).
          Be aware that you should add *at least* the TYPO3 and PHP version constraints
          to this file to make sure everything is working properly.
 
+         For legacy non-Composer installation the :file:`ext_emconf.php` file
```
Comment (mbrodala) [inline Documentation/ExtensionArchitecture/DeclarationFile/Index.rst:190]:
```suggestion
         For Legacy installations the :file:`ext_emconf.php` file
```

File: Documentation/ExtensionArchitecture/FilesAndLocations/Index.rst [hunk 20]
```diff
@@ -35,11 +35,17 @@ use the content for specific functionality. For example, if a svg logo of your e
 is placed at :file:`Resources/Public/Icons/Extension.svg`, the Extension Manager
 will show that image.
 
-Most of these files are not required. The exception is :file:`ext_emconf.php`:
-You can not have a TYPO3
-extension recognized by TYPO3 without this file.
+Most of these files are not required. The exception are :file:`ext_emconf.php`
+in legacy (non-Composer) installations and :file:`composer.json` in
```
Comment (mbrodala) [inline Documentation/ExtensionArchitecture/FilesAndLocations/Index.rst:39]:
```suggestion
in Legacy installations and :file:`composer.json` in
```

File: Documentation/ExtensionArchitecture/FilesAndLocations/Index.rst [hunk 21]
```diff
@@ -35,11 +35,17 @@ use the content for specific functionality. For example, if a svg logo of your e
...
+Composer-based installations.
```
Comment (mbrodala) [inline Documentation/ExtensionArchitecture/FilesAndLocations/Index.rst:40]:
```suggestion
Composer installations.
```

File: Documentation/ExtensionArchitecture/FilesAndLocations/Index.rst [hunk 22]
```diff
@@ -35,11 +35,17 @@ use the content for specific functionality. For example, if a svg logo of your e
...
 
-In general, do not introduce your own files in the root directory of
+.. note::
+   It is recommended to keep :file:`ext_emconf.php` and :file:`composer.json` in
+   any public extension that is published to TYPO3 Extension Repository (TER), and
+   to ensure optimal compatibility with Composer-based installations and legacy
+   mode.
```
Comment (mbrodala) [inline Documentation/ExtensionArchitecture/FilesAndLocations/Index.rst:46]:
```suggestion
   to ensure optimal compatibility with Composer and Legacy installations.
```

File: Documentation/ExtensionArchitecture/FilesAndLocations/Index.rst [hunk 23]
```diff
@@ -48,29 +54,39 @@ extensions with the name prefix :file:`ext_`, because that is reserved.
 :file:`composer.json`
 ---------------------
 
-*-- required*
+*-- required* in composer-based installations
```
Comment (mbrodala) [inline Documentation/ExtensionArchitecture/FilesAndLocations/Index.rst:57]:
```suggestion
*-- required* in Composer installations
```

File: Documentation/ExtensionArchitecture/FilesAndLocations/Index.rst [hunk 24]
```diff
@@ -48,29 +54,39 @@ extensions with the name prefix :file:`ext_`, because that is reserved.
...
 
 For more information, see :ref:`composer-json`.
 
+.. versionchanged:: 11.4
+   The ordering of installed extensions and their dependencies are loaded from
+   the :file:`composer.json` file, instead of :file:`ext_emconf.php` in
+   composer-based installations.
```
Comment (mbrodala) [inline Documentation/ExtensionArchitecture/FilesAndLocations/Index.rst:64]:
```suggestion
   Composer installations.
```

File: Documentation/ExtensionArchitecture/FilesAndLocations/Index.rst [hunk 25]
```diff
@@ -48,29 +54,39 @@ extensions with the name prefix :file:`ext_`, because that is reserved.
...
 .. _ext_emconf-php:
 
 :file:`ext_emconf.php`
 ----------------------
 
-*-- required*
-
-Definition of extension properties. This is the only mandatory file in the extension.
-It describes the extension.
+*-- required* in legacy installations
```
Comment (mbrodala) [inline Documentation/ExtensionArchitecture/FilesAndLocations/Index.rst:79]:
```suggestion
*-- required* in Legacy installations
```

File: Documentation/ExtensionArchitecture/FilesAndLocations/Index.rst [hunk 26]
```diff
@@ -48,29 +54,39 @@ extensions with the name prefix :file:`ext_`, because that is reserved.
...
 
-Name, category, status etc. are used by the Extension Manager. The content of this file
-is described in more details in :ref:`extension-declaration`. Note
-that it is auto-written by the Extension Manager when extensions are imported from the repository.
+Definition of extension properties.
 
-.. note::
+Name, category, status etc. are used by the Extension Manager in legacy
```
Comment (mbrodala) [inline Documentation/ExtensionArchitecture/FilesAndLocations/Index.rst:83]:
```suggestion
Name, category, status etc. are used by the :guilabel:`Extensions` module in Legacy
```

File: Documentation/ExtensionArchitecture/FilesAndLocations/Index.rst [hunk 27]
```diff
@@ -48,29 +54,39 @@ extensions with the name prefix :file:`ext_`, because that is reserved.
...
+installations. The content of this file is described in more details in
+:ref:`extension-declaration`.
 
-   If this file is *not* present, the Extension Manager will *not* find the
-   extension.
+For legacy non-Composer installation the :file:`ext_emconf.php` file is the
```
Comment (mbrodala) [inline Documentation/ExtensionArchitecture/FilesAndLocations/Index.rst:87]:
```suggestion
For Legacy installations the :file:`ext_emconf.php` file is the
```

**Review by mbrodala — CHANGES_REQUESTED** [review]
Maybe we should also add a link to these pages at the top:

- https://docs.typo3.org/m/typo3/tutorial-getting-started/master/en-us/Installation/Install.html
- https://docs.typo3.org/m/typo3/tutorial-getting-started/master/en-us/Installation/LegacyInstallation.html

This way we don't have to clarify the facts over and over again.

File: Documentation/ApiOverview/Configuration/ConfigurationOverview.rst (hunk 3 above)
Comment (linawolf) [reply Documentation/ApiOverview/Configuration/ConfigurationOverview.rst:46]: I would keep Composer-based installation as a composer installation might be mistaken for the installation of composer (not with composer)

File: Documentation/ApiOverview/Configuration/ConfigurationOverview.rst (hunk 4 above)
Comment (linawolf) [reply Documentation/ApiOverview/Configuration/ConfigurationOverview.rst:49]: Legacy is not a proper noun, it is just an adjective

File: Documentation/ExtensionArchitecture/ComposerJson/Index.rst (hunk 5 above)
Comment (linawolf) [reply Documentation/ExtensionArchitecture/ComposerJson/Index.rst:12]: see above

File: Documentation/ExtensionArchitecture/ComposerJson/Index.rst (hunk 6 above)
Comment (linawolf) [reply Documentation/ExtensionArchitecture/ComposerJson/Index.rst:17]: I would keep this as it makes more clear what kind of installation is talked about

File: Documentation/ExtensionArchitecture/ComposerJson/Index.rst (hunk 7 above)
Comment (linawolf) [reply Documentation/ExtensionArchitecture/ComposerJson/Index.rst:82]: see above

File: Documentation/ExtensionArchitecture/ComposerJson/Index.rst (hunk 8 above)
Comment (linawolf) [reply Documentation/ExtensionArchitecture/ComposerJson/Index.rst:203]: You might not depend on other system extensions.

File: Documentation/ExtensionArchitecture/FilesAndLocations/Index.rst (hunk 23 above)
Comment (linawolf) [reply Documentation/ExtensionArchitecture/FilesAndLocations/Index.rst:57]: see above

File: Documentation/ExtensionArchitecture/FilesAndLocations/Index.rst (hunk 24 above)
Comment (linawolf) [reply Documentation/ExtensionArchitecture/FilesAndLocations/Index.rst:64]: see above

File: Documentation/ExtensionArchitecture/FilesAndLocations/Index.rst (hunk 25 above)
Comment (linawolf) [reply Documentation/ExtensionArchitecture/FilesAndLocations/Index.rst:79]: see above

File: Documentation/ExtensionArchitecture/FilesAndLocations/Index.rst (hunk 27 above)
Comment (linawolf) [reply Documentation/ExtensionArchitecture/FilesAndLocations/Index.rst:87]: see above

**Comment by linawolf** [comment]
@mbrodala please have another look

File: Documentation/ApiOverview/Configuration/ConfigurationOverview.rst (hunk 3 above)
Comment (mbrodala) [inline Documentation/ApiOverview/Configuration/ConfigurationOverview.rst:46]:
```suggestion
   Composer configuration, required in Composer-based installations
```

**Review by mbrodala — CHANGES_REQUESTED** [review]

**Comment by mbrodala** [comment]
Please also consider my remaining suggestions.

**Comment by linawolf** [comment]
We are mixing up things way to much here. This ticket is merely about Important: #95647. You can open another ticket and make more changes to the wordings of Classic mode etc that where not introduced with this ticket

**Review by mbrodala — APPROVED** [review]

**Pull request merged**